src/cpp/session/SessionSuspend.cpp

```diff
@@ -122,6 +427,73 @@ void suspendIfRequested(const boost::function<bool()>& allowSuspend)
       // errors will be logged/reported internally and we will move on
       suspendSession(false);
    }
+
+   // timeout suspend
+   checkForTimeout(allowSuspend);
+}
+
+std::string mostSignificantTimeAgo(boost::posix_time::ptime before, boost::posix_time::ptime after = boost::posix_time::second_clock::universal_time())
+{
+   if (before == boost::posix_time::not_a_date_time || after == boost::posix_time::not_a_date_time)
+      return "";
+
+   boost::posix_time::time_duration diff = after - before;
+   if (diff.is_negative())
+   {
+       diff.invert_sign();
+   }
+
+   // determine the most significant non-zero time resolution
+   if (diff < boost::posix_time::minutes(2))
+      return ""; // Stay quiet if not enough time has passed
+   if (diff < boost::posix_time::hours(1))
+      return std::to_string(diff.minutes()) + " minutes ago";
+   if (diff < boost::posix_time::hours(24))
+      return std::to_string(diff.hours()) + " hour(s) ago";
+   if (diff < boost::posix_time::hours(24 * 365))
+      return std::to_string(diff.hours() / 24l) + " day(s) ago";
+   else
+      return std::to_string(diff.hours() / (24l * 365l)) + " year(s) ago";
+}
+
+std::string getResumedMessage()
+{
+   if (s_timeoutState == kWaitingForInactivity)
+   {
+      std::string xAmountOfTimeAgo = mostSignificantTimeAgo(s_blockingTimestamp);
+      if (xAmountOfTimeAgo.empty())
+         return "";
+
+      std::string blockedTime = boost::posix_time::to_simple_string(s_blockingTimestamp);
+      return "Welcome Back! Session has been running in the background since "
+             + blockedTime
+             + " UTC ("
+             + xAmountOfTimeAgo
+             + ")\n";
+   }
+   else
+   {
+      boost::posix_time::ptime suspensionTimestamp = module_context::activeSession().suspensionTime();
+      std::string xAmountOfTimeAgo = mostSignificantTimeAgo(suspensionTimestamp);
+      if (xAmountOfTimeAgo.empty())
+         return "";
+
+      std::string suspendedTime = boost::posix_time::to_simple_string(suspensionTimestamp);
+      return "Session restored from your saved work on "
```

Comment:
I'm worried that these will wind up being just a little noisy in the R console, but we can get them in as is and see how it feels in real-world usage.

Reply:
I thought the same, and have it configured to at least not display the message when they just refresh the page. Hopefully that reduces any annoying repeats